drivers/hmis/app/models/hmis/workflow_definition/graph.rb

```diff
@@ -28,14 +28,18 @@ def walk(entrypoint_ids: nil, stop_when: nil)
           next if visited.include?(node)
 
           yielder << node unless node.id.in?(skip_ids)
-          visited.add(node) unless node.id.in?(skip_ids) # don't return entrypoints if from args
+          visited.add(node) unless node.id.in?(skip_ids) # if entrypoints were passed as arguments, don't return them
 
-          # careful, this logic is a bit ugly/fragile
-          next unless node.id.in?(skip_ids) || (stop_when.nil? || !stop_when.call(node))
+          # If this node was passed as an entrypoint, don't check the stop_when condition. (Otherwise we wouldn't traverse any nodes)
```

Comment:
@ttoomey I wanted to take a stab at clarifying this logic, since every time I looked at it I started getting confused ("wait, why do we want to `next` if this _isn't_ one of the IDs that we should skip?"). I had also contributed to the pile of double-negatives here in [this PR](https://github.com/greenriver/hmis-warehouse/pull/5204/files#diff-c5787306c74350094172ea3657b96a8e1c20931a05d675d29a34b71769be0b9eR34-R38). I've tried to just clarify, without actually changing the logic. And I added spec tests below to convince myself of the behavior. Let me know if this looks okay to you. 🙏

Reply:
@martha thanks for improving this! It looks great to me, I think it should have the same behavior but is easier to read